File: HW/HW1/PalindromeChecker.py
import re

def isPalindrome(s: str) -> bool:
    s = re.sub(r'[^a-zA-Z0-9]', '', s).lower()

    left, right = 0, len(s) - 1
    
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
        
    return True

print(isPalindrome("ab:b,a"))
print(isPalindrome("race a car"))

File: HW/HW1/PalindromeChecker.java
package HW1;

public class PalindromeChecker {
    public boolean isPalindrome(String s) {
        s = s.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        
        int left = 0, right = s.length() - 1;
        
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        
        return true;
    }

    public static void main(String[] args) {
        PalindromeChecker checker = new PalindromeChecker();

        System.out.println(checker.isPalindrome("ab:b,a"));
        System.out.println(checker.isPalindrome("race a car"));
    }
}